src/awkward1/operations/convert.py

```diff
@@ -1139,6 +1139,180 @@ def getfunction(layout, depth):
+            if mask is not None:
+                return recurse(layout.broadcast_tooffsets64(layout.compact_offsets64()), mask)
+            else:
+                return recurse(layout.broadcast_tooffsets64(layout.compact_offsets64()))
+
+        elif isinstance(layout, (awkward1.layout.RecordArray)):
+            values = [recurse(x) for x in layout.contents]
+            # Arrow rejects the values list if every child array is not of same length
+            min_list_len = min(map(len, values))
+
+            if mask is not None:
+                for i in range(len(values)):
+                    values[i] = values[i].slice(0, len(mask)) 
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)   
+                return pyarrow.StructArray.from_arrays(values, layout.keys()).take(pyarrow.array(indices))
+            else:
+                for i in range(len(values)):
+                    values[i] = values[i].slice(0, min_list_len)
+                return pyarrow.StructArray.from_arrays(values, layout.keys())
+        
+        elif isinstance(layout, (awkward1.layout.UnionArray8_32, 
+                                awkward1.layout.UnionArray8_64, 
+                                awkward1.layout.UnionArray8_U32)):
+        
+            values = [recurse(x) for x in layout.contents]
+            if mask is not None:
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)
+                return pyarrow.UnionArray.from_dense(recurse(layout.tags), recurse(layout.index), values).take(pyarrow.array(indices))
+            return pyarrow.UnionArray.from_dense(recurse(layout.tags), recurse(layout.index), values)
+        
+        elif isinstance(layout, (awkward1.layout.IndexedArray32, 
+                                awkward1.layout.IndexedArray64, 
+                                awkward1.layout.IndexedArrayU32)):
+            if mask is not None:
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)
+                return pyarrow.DictionaryArray.from_arrays(recurse(layout.index), recurse(layout.content)).take(pyarrow.array(indices))
+            else:
+                return pyarrow.DictionaryArray.from_arrays(recurse(layout.index), recurse(layout.content))
+
+        elif isinstance(layout, (awkward1.layout.IndexedOptionArray32, 
+                                awkward1.layout.IndexedOptionArray64)):
+            # Convert this to a Union Array
+            dummy_numeric_val = awkward1.Array([]).layout
```

Comment:
I might need some opinion on this, as to what the dummy value should be. According to Awkward, it should be the first element of the `IndexedOptionArray`. Should we go with this?

Reply:
We'll be in trouble if there's ever an array consisting entirely of missing values, but we can come back to that problem later. Sure, the first entry (assuming it exists) is as good a choice as any. Unfortunately, it needs to be copy-pasted under all of the None values.

Reply:
Could we perhaps proceed with the conversion to UnionArray first as you suggested, the difference being that in `tag0`  it could be the first element of the `IndexedOptionArray`  in question and `tag1` is the IndexedOptionArray itself?